components/num-util/src/fixed_decimal.rs

```diff
@@ -373,6 +377,135 @@ impl fmt::Display for FixedDecimal {
+        let no_sign_str: &str;
+        let mut is_negative = false;
+        if &input_str[0..1] == "-" {
+            is_negative = true;
+            no_sign_str = &input_str[1..];
+        } else {
+            no_sign_str = input_str;
+        }
+        // Compute length of each string once and store it, so if you use that multiple times,
+        //  you don't compute it multiple times
+        // has_dot: shows if your input has dot in it
+        // dot_index: gives the index of dot (after removing the sign)
+        let no_sign_str_len = no_sign_str.chars().count();
+        let mut has_dot = false;
+        let mut dot_index = no_sign_str_len;
+        // The following loop computes has_dot, dot_index, and also checks to see if all
+        // characters are digits and if you have at most one dot
+        // Note: Input of format 111_123 is detected as syntax error here
+        // Note: Input starting or ending with a dot is detected as syntax error here (Ex: .123, 123.)
+        for (i, c) in no_sign_str.chars().enumerate() {
+            if c == '.' {
+                match has_dot {
+                    false => {
+                        dot_index = i;
+                        has_dot = true;
+                        if i == 0 || i == no_sign_str.len() - 1 {
+                            return Err(Error::Syntax);
+                        }
+                    }
+                    true => {
+                        return Err(Error::Syntax);
+                    }
+                }
+            } else if !c.is_digit(10) {
+                return Err(Error::Syntax);
+            }
+        }
+        // no_dot_str: shows the string when the dot is removed from it
+        let no_dot_str = no_sign_str.replace(".", "");
+        let no_dot_str_len = no_dot_str.chars().count();
+
+        // left_zeros: number of zeros at the left of no_dot_str
+        // right_zeros: number of zeros at the right of no_dot_str
+        // These are not necessarily leading or trailing zeros
+        // Example:
+        //     input string     left_zeros      right_zeros
+        //     00123000             2                3
+        //     0.0123000            2               3
+        //     001.23000            2               3
+        //     001230.00            2               3
+        let mut left_zeros = 0;
+        for (_i, c) in no_dot_str.chars().enumerate() {
+            if c == '0' {
+                left_zeros += 1;
+            } else {
+                break;
+            }
+        }
+        // If the input only has 0 (like 000, 00.0, -00.000) we handle the situation here
+        // and return the correct FixedDecimal now and don't run the rest of the code
+        if left_zeros == no_dot_str_len {
+            let mut dec: FixedDecimal = 0.into();
+            dec.is_negative = is_negative;
+            dec.upper_magnitude = left_zeros as i16 - 1i16;
```

Comment:
Casting from `usize` to `i16` is lossy.  What if you have a string with 100,000 zeros?

Try this:

```rust
dec.upper_magnitude = (dot_index - 1).try_into()?;
dec.lower_magnitude = -((no_dot_str_len - dot_index).try_into()?);
```

Note: both of these operations depend on the fact that `dot_index` is >= 1.

Please write test cases that test the extremes (strings of all zeros that are slightly shorter, slightly longer, and exactly the right length).  The maximum string length, of course, not including the `-` or the `.`, is u16::MAX.

Reply:
Actually, I tried `try_into()?` but I get an error saying that `try_into()` is not defined for `usize`. Also, if I have a string with more than `i16::MAX' before or after the floating-point,  isn't that a Limit error? 
I like to compare `(dot_index - 1)` with `std::i16::MAX` and if it is smaller then set `dec.upper_magnitude` equal to that, and return `Err(Error::Limit)` otherwise. But I have trouble comparing these quantities bc one of them is `usize` and the other one is `isize`. (I have the same problem in many other occasions too, and didn't find a good answer on the Internet)

Reply:
For upper_magnitude, you could use `i16::try_from(dot_index - 1)`.  You'll need to add a mapping for the error to the Error type that results in Error::Limit, such as

```rust
dec.upper_magnitude = i16::try_from(dot_index - 1).map_err(|e| Error::Limit)?;
```

Alternatively, you could write it out explicitly:

```rust
let temp_upper_magnitude = dot_index - 1;
if temp_upper_magnitude > i16::MAX as usize {
    return Err(Error::Limit);
}
dec.upper_magnitude = temp_upper_magnitude as i16;
```

For lower_magnitude, you're right about i16::MIN.  It's a little harder to read, but try

```rust
let temp_lower_magnitude = no_dot_str_len - dot_index;
// Note: ((i16::MIN as u16) as usize) == 32768
if (temp_lower_magnitude > (i16::MIN as u16) as usize {
    return Err(Error::Limit);
}
dec.lower_magnitude = (temp_lower_magnitude as i16).wrapping_neg();
```

Play around with it and see if you can get something more readable. :)